### ISSUE TITLE
vector outputs don't take start offset and gaps into account

### ISSUE THREAD

**Issue by cannam** [piper-audio/ugly-duckling#8]
A Vamp plugin can have a fixed-sample-rate output without actually supplying any features for every moment in the output's sampled timeline. For example, pYin and the cepstral pitchtracker both use a fixed-sample-rate output for pitch tracks but emit no features for any times where no pitched audio was detected.

At the moment the app is displaying these as a single trace, all bunched up and starting from time 0, without any gaps in it. It should probably display a separate trace for each set of contiguous points. The fault is probably in the vector encoding in the higher-level API, which does not return the start time and has no way to indicate a gap (it should perhaps return more than one separate vector).

(The Python Vamp host module's higher-level API has exactly the same problem.)

**Comment by cannam** [comment]
This is a theoretical problem with the matrix output shape as well. (Matrix outputs seldom have gaps in reality, but they do sometimes have non-zero start times.)

Essentially, anything having a OneSamplePerStep output can be deemed to consist of a single vector or matrix starting at time zero (or, more precisely, at the start time of the input audio). But a FixedSampleRate output is different. There's no obligation for it to actually produce a feature at any given moment on its sample timeline. So although it _probably_ will have a continuous series of samples that can be represented as a single vector or matrix, there's a chance it will have two or more series with gaps between them. In a pathological case it could have a feature for every second sample, meaning half as many individual vectors or matrices as there are sample points.

**Comment by cannam** [comment]
Note the Vamp Test Plugin's "FixedSampleRate/Timed" output (curve-fsr-timed) provides a test for this, in that there are gaps in the output. There isn't a case to cover discontinuous matrices though, or non-zero start times.

**Comment by LucasThompson** [comment]
So, FixedSampleRate feature outputs aren't necessarily fixed spaced? Well, I'll be damned! 

I guess this issue also applies to VariableSampleRate outputs?

I'm wondering about ways to encode the information, your suggestion in the first comment about having more than one separate vector makes sense. Though, I'm wondering, is the spacing between the vectors implicit? Wouldn't each vector need a timestamp for the start time? Then we kind of move back towards the traditional (or "list") output encoding.

**Comment by cannam** [comment]
VariableSampleRate outputs are not encoded as vectors or matrices, only as lists (with a timestamp on each feature). So there's no issue there.

FixedSampleRate outputs have a fixed-spacing sample timeline. Every feature's timestamp has to be on that fixed sample grid (and if it isn't, the host has to round the timestamp so that it is). But there doesn't have to be a feature for every space on the sample grid, and there could even be more than one for some.

If we did start a new vector/matrix each time we found a gap (or more precisely, each time we found a feature whose timestamp was not exactly one sample space further on than the previous one) then yes, each vector would need a start time. Since the most common case is to have a single continuous vector/matrix, that means most of the time we'd have an array containing one vector plus its start time, which most often would be time 0.

**Comment by LucasThompson** [comment]
Ah, that was silly of me there regarding the VariableSampleRate outputs. 

So, as it stands, these are the current types for representing the higher level ("collected") features:

```TypeScript
type FeatureCollectionShape = "matrix" | "vector" | "list";

interface FeatureCollection {
  shape: FeatureCollectionShape;
  data: FeatureList | Float32Array | Float32Array[];
}

interface FixedSpaceFeatures extends FeatureCollection {
  stepDuration: number;
}
```

They're not well thought out, and pretty much the first thing that I dumped in my editor. There are issues with the fact that the union types for the shape and the data are actually paired together (`"matrix"` with `Float32Array[]`, `"vector"` with `Float32Array` and `"list"` with `FeatureList`) and I've not put any time into thinking of a way to leverage the type system / compiler to ensure type safety there. 

In previous comments, we are describing something like the following for the outputs with gaps:

```TypeScript
type Matrix = Float32Array[];
type Vector = Float32Array;

/*TODO think of a better name prefix, SparseFixedSampleRate*? hmmm...*/
interface GappyFeatureData<T extends Vector | Matrix> {
  startTime: Timestamp;
  data: T;
}

```

Which might require altering FeatureCollection to something like:

```TypeScript
type FeatureCollectionData = FeatureList 
  | Vector 
  | Matrix 
  | GappyFeatureData<Vector>[] 
  | GappyFeatureData<Matrix>[];

interface FeatureCollection<T extends FeatureCollectionData> {
  shape: FeatureCollectionShape;
  data: T;
}
```

The use of generics for `GappyFeatureData` might be redundant, and changing `FeatureCollection` to a generic type would involve some refactoring.

**Comment by cannam** [comment]
I mentioned that the Python host has the same problem. A related question therefore is, how can one change the structures used in that host so as to fix this, and is it possible to do that without breaking existing code? If so, there might be good reason to adopt the same solution here.

The Python host returns a structure of the form

`{ "matrix": ( 0.1, np.array([[1,2,3], [4,5,6]], dtype=float32) ) }`

which is then typically unpacked like

`stepsize, chromadata = result["matrix"]`

The stepsize isn't labelled - it's just the first element in a member tuple. So we can't simply add a start time to the same structure without making it really obscure to interpret. And there is only room for one array, supplying a list of arrays would change the unpacking behaviour.

An analogous structure for multiple matrices would be surprisingly complicated even if we weren't trying to retain compatibility. For example

```
{ "matrix": { "stepSize": 0.1, [ { "startTime": 0.0, "data": np.array([[1,2,3], [4,5,6]], dtype=float32) },
                                 { "startTime": 1.0, "data": np.array([[3,2,1], [5,4,6]], dtype=float32) } ] } }
```
or a simpler structure but with some redundancy for the step size,

```
{ "matrix": [ { "stepSize": 0.1, "startTime": 0.0, "data": np.array([[1,2,3], [4,5,6]], dtype=float32) },
              { "stepSize": 0.1, "startTime": 1.0, "data": np.array([[3,2,1], [5,4,6]], dtype=float32) } ] }
```
To do this while retaining compatibility, one could

 * have a separate output shape for multiple matrices (and so on for multiple vectors), e.g. "matrices", with the above form - but including it only if there are multiple matrices, using the existing matrix structure where there is only a single one. That would require going through all the features before knowing what shape they were though. Or
 * have a separate field for additional matrices beyond the first one - or even for the first matrix if its start time is non-zero? i.e. something like

```
{ "matrix": ( 0.1, np.array([[1,2,3], [4,5,6]], dtype=float32) ),
  "extraMatrices": [ { "stepSize": 0.1, "startTime": 1.0, "data": np.array([[3,2,1], [5,4,6]], dtype=float32) } ] }
```
This is horribly uneven in structure though. Or
 * have a separate field for individual matrices even if there is only one - in which case effectively duplicating the entire structure, though presumably the underlying arrays will be shared so long as they're just in-memory. That is

```
{ "matrix": ( 0.1, np.array([[1,2,3], [4,5,6]], dtype=float32) ),
  "allMatrices": [ { "stepSize": 0.1, "startTime": 0.0, "data": np.array([[1,2,3], [4,5,6]], dtype=float32) },
                   { "stepSize": 0.1, "startTime": 1.0, "data": np.array([[3,2,1], [5,4,6]], dtype=float32) } ] }
```

**Comment by cannam** [comment]
Another idea (to make it easier for very naive client code to get something meaningful out of this) might be to fill in the missing matrix (or vector) cells to make a single, coherent structure, and then provide individual matrices separately. In Python terms:

```
{ "matrix": ( 0.1, np.array([[1,2,3], [4,5,6], [0,0,0], [0,0,0], [0,0,0], [0,0,0], [0,0,0], [0,0,0], [0,0,0], [0,0,0], [3,2,1], [5,4,6]], dtype=float32) ),
  "individualMatrices": [ { "stepSize": 0.1, "startTime": 0.0, "data": np.array([[1,2,3], [4,5,6]], dtype=float32) },
                          { "stepSize": 0.1, "startTime": 1.0, "data": np.array([[3,2,1], [5,4,6]], dtype=float32) } ] }
```
Disadvantages: 

* Potentially involves a lot more memory/data - though in the "normal" case with just a single shape, the matrix and the first individual matrix would be identical and the backing array could still be shared
* Assumes presence of a meaningful null element that can be used to fill in the gaps (I used 0 above, but why?)
* Having two "views" of the same data creates ambiguity about how to interpret it
* "individualMatrices" is a _very_ awkward name

Advantages:

* Backward compatible
* Gets you something correct and visually meaningful even if you just read the single shape and blat it on the screen or whatever
* Eliminates risk (that some of the earlier-mentioned formats have) that one might read only the first shape and never realise there are any more of them
* A version of the data with the gaps filled in is potentially useful anyway for some purposes (e.g. it's the format you want for your MIREX pitch-tracking submission)

**Comment by LucasThompson** [comment]
The last solution seems pretty sensible, especially for backward compatibility in the python case. 

In the case of piper, though, the sharing of the data is perhaps less plausible. If we think about the fact the extraction happens on a server which is perhaps remote, in a separate process without shared memory etc, the serialised response payloads inevitably end up having to contain a lot more data. 

Unless the responses just contain the "filled" matrices and the piper client is left to construct the individual matrices from the potentially ambiguous null value, which presumably will change depending on the feature. Or the encoding is such that data for indexing the "filled" matrix is provided.

**Comment by cannam** [comment]
You're right, in Piper-land we don't want to send everything twice.

But we should aim to send "correct" data (even if it does in general need more fiddling with by the recipient) rather than have occasional situations where you have to deduce something from incomplete information. This suggests that we can't avoid sending the individual matrices, at least where they differ from the single matrix.

So while that last suggestion might work in the Python host or in client code generally, for the protocol we have to do one of the following

* send either the single matrix or multiple matrices (depending on whether multiple ones are necessary or not) - i.e. make single and multiple be effectively different output shapes. Or
* always send an array of matrices, but having only one element if only one was needed (i.e. most of the time).

And similarly for vectors of course.

**Comment by LucasThompson** [comment]
I'm leaning towards the separate output shape option - would this constitute the "tracks" shape from [our proposed output shapes](https://code.soundsoftware.ac.uk/projects/vamp-js/wiki/Proposed_Output_Shapes)?

**Comment by cannam** [comment]
Interesting. Different use case (that was intended for simultaneous tracks, while this was provoked by failure to handle consecutive separate tracks) but I think it might be the same structure.

The equivalent situation with the matrix output only really arises by symmetry - I don't recall any plugin ever actually wanting to send multiple disconnected matrices from an output. This does suggest that we might be able to get away with sending only a single one and fudging it, _except_ that we do also want to carry the start time of matrix even where there is only one. In fact in general we need to remember to add the start time regardless of whether we're sending multiple tracks or only a single one.

**Comment by cannam** [comment]
How about this:

### Vectors and tracks

For fixed-sample-rate vector outputs (those known to have exactly one value per feature):

- If the output features start at time zero and have no gaps, return a "vector" shape with a single array of values and a step time
- In any other case (gaps or non-zero start time), return a "tracks" shape with an array of { startTime, stepDuration, values } structures, one for each contiguous series of features in the output

It's not obvious whether that's the right thing for the case where there are no gaps but the vector starts somewhere other than time 0. We could make that a vector shape, in which case we'll need the vector to have a start time as well as a step duration (leading to complications providing backward compatibility in e.g. the Python case), or we could make it a tracks shape as above. Any preference?

This also has the disadvantage that we can't determine the shape from output static data alone, only after we've seen all the features.

(Of course, one-sample-per-step vectors remain vectors.)

### Matrices

Given that we have no examples yet of plugins wanting to return multiple regions of matrices, let's just stick to a single matrix shape for now. That is:

- If the output features start at time zero and have no gaps, return a "matrix" shape with a single 2d array of values and a step time
- If the output features start at some other time and/or have any gaps, fill in the gaps with 0 values and still return a "matrix" shape.

If at any point it becomes obvious that we really wanted a matrices or grids shape, we can always add that separately (similarly to the way our tracks shape is separate from the vector shape).

### Other uses of the tracks shape

The "tracks" shape was initially suggested for multiple simultaneous tracks, whereas the above was prompted by problems handling multiple consecutive but discontiguous tracks.

It's sort-of possible to represent simultaneous tracks using a one-valued fixed-rate feature, if your plugin has calculated the whole time range in advance - just emit one whole track and then go back to the same start time for the next track's features. That would work OK in this scheme.

However, that isn't always possible or suitable (for example it can't work causally), and some existing plugins (e.g. the CHP plugin used by Tony) instead represent simultaneous tracks as fixed-rate features with varying numbers of values. Unfortunately there's no reliable way to deduce this structure properly without more information or convention - e.g. how do you know which tracks are associated with which values in the features? This may be a case where the plugin needs to be more explicit than is possible with Vamp.

**Comment by cannam** [comment]
(I am currently having a crack at implementing this in the Python code. It's really quite awkward not to know which shape you want until you've seen all the features - I think this is always going to be a tricky situation here.)

**Comment by LucasThompson** [comment]
Some thoughts and questions - not too much to add, happy to see how the python stuff goes and then discuss from there where to go with the piper-js stuff:

### Vectors and tracks
- I think it makes sense in general to have a new shape which represents a collection of other shapes, which are not necessarily contiguous. So it follows that the start time would be a property of the elements in that collection shape 
- After a small amount of time pondering, I don't really mind the consequence of having a track with a single element just to indicate a non-zero start time for a single vector.

### Matrices
- My initial thoughts here are that it might not really be much more effort to just do the same as above for matrices vs filling in the gaps with zeros, even if we can't think why a plugin would produce that output? I guess even if that were the case it doesn't make it a sensible thing to do.

**Comment by cannam** [comment]
Now that this is working OK in the Python, I'm taking another look at it in HigherLevelUtilities.ts as well (in my own fork for now).

The structure currently used in the JS/TS code has a string tag for the type of shape, and then a union type for the data. A problem with it is that the static data type does not make explicit the connection between type of shape and type of data (e.g. the compiler can't check that you always use the Float32Array data if you happen to have a vector type).

The data structure we use in the Python is different because it makes the type of data implicit on the fields present in the structure (duck-typingly I spose). I guess that looks a bit like this, in Typescript terms:

```TypeScript
interface VectorFeature {
  stepDuration: number; /* or Timestamp - we are inconsistent about this I think? */
  data: Float32Array;
}

interface MatrixFeature {
  stepDuration: number; /* or Timestamp - we are inconsistent about this I think? */
  data: Float32Array[];
}

interface TrackFeature {
  startTime: number; /* or Timestamp */
  stepDuration: number; /* or Timestamp */
  data: Float32Array;
}

interface FeatureCollection {
  vector?: VectorFeature;
  matrix?: MatrixFeature;
  tracks?: TrackFeature[];
  list?: FeatureList;
}
```

That would ensure (if my syntax works out) that you always access the VectorFeature with the vector shape, etc., but it has other problems:

* nothing in the FeatureCollection enforces the fact that we can only have one shape in there (it's theoretically possible one could receive a FeatureCollection with both matrix and list, say)
* it would leave you checking the presence/absence of each field in order to find out what the shape is (you could add a shape tag to FeatureCollection but then you have the first problem back again)

Presumably the solution to the first of those problems would be to use an explicit union type, but I'm not totally clear what that looks like in terms of both declaration & dereference.

I'm also interested in what this all translates to in Javascript (as opposed to Typescript) terms -- the code listed above I think translates to something basically the same as the Python version, except that the Python perhaps ill-advisedly uses tuples for VectorFeature and MatrixFeature.

**Comment by cannam** [comment]
At the moment I'm trying something like this.

```TypeScript
export interface VectorFeatures {
    stepDuration: number;
    data: Float32Array;
}

export interface MatrixFeatures {
    stepDuration: number;
    data: Float32Array[];
}

export interface TrackFeature {
    startTime: number;
    stepDuration: number;
    data: Float32Array;
}

export type TrackFeatures = TrackFeature[];

export type FeatureCollectionShape = "matrix" | "vector" | "tracks" | "list";

export type FeatureCollection = {
    shape: FeatureCollectionShape;
    features: VectorFeatures | MatrixFeatures | TrackFeatures | FeatureList;
}
```

**Comment by cannam** [comment]
I have the above in the "dev-tracks" branch of cannam/piper-js, with one change:

```TypeScript
export type FeatureCollection = {
    shape: FeatureCollectionShape;
    collected: VectorFeatures | MatrixFeatures | TrackFeatures | FeatureList;
}
```

i.e. changing "features" to "collected". I'm not sure that's for the best, but "features" isn't a good name either, because we are already using it in SimpleResponse, so we'd end up with features.features all over the place.

I haven't attempted to implement the actual "tracks" reshaping yet.

**Comment by cannam** [comment]
Today's plan is to take advantage of the fact that the Vamp Test Plugin's existing curve-fsr-timed output should have "tracks" shape, and to flesh out the (ancient stub of a) VampTestPluginTest in piper-js/test so that it can correctly fail a test for this output. Then write the reshape code to match.

**Comment by LucasThompson** [comment]
TDD Wednesday

**Comment by cannam** [comment]
So, I got a bit stuck yesterday afternoon. I did what I described in the comment above: I updated the HigherLevelUtilities reshape code so that it reshaped into tracks as appropriate, passing a test using the Vamp Test Plugin.

The problem I then faced was how to get this shape recognised in ugly-duckling, and here I got stuck in the StreamingService and FeatureReducers logic. It's not clear to me where the reshape is supposed to be happening in this setup. I understand that some of the code there derives from a (postponed) effort to add incremental rendering, which means the server is sending some sort of response payload for each input process block. I don't really understand what is expected to be in the response payloads. There are several possibilities for what's happening here:

1. The client is expecting to do the reshape after getting all the responses. The response payloads consist of raw features, so effectively always "list" shape (even though the FeatureReducers logic seems to be inspecting structures of FeatureCollection type, which ought to mean post-reshape). _or_

2. The server has already begun reshaping before it sends any of these responses, and is effectively tearing apart the collected shape in order to send it a bit at a time. This is what might happen if you did the reshape based on static metadata and then produced the reshaped data lazily using a generator mechanism. The client gets a series of FeatureCollection objects of the correct shape, but it needs to squish them into one big FeatureCollection object (making the assumption e.g. that a series of vector shapes actually represents one long vector).  _or_

3. The server reshapes when it has all the features available, and returns a single FeatureCollection response at that time. So the FeatureReducers logic only actually applies once, reducing a single FeatureCollection into itself. Meanwhile progress updating is based on the number of requests sent to the server, and this happens somewhere else that I haven't found yet. _or_

4. The server reshapes when it has all the features available, but in order to make progress updates available to the client, it first returns a trivial FeatureCollection response to each of the input requests so that the client can use those for progress updating. This option has sub-options 4a and 4b, depending on whether the server also splits up the final shape à la option 2 above or returns it in a single collection à la option 3.

I think that the answer is probably 4b, but I'm not certain. It doesn't help that I'm already a bit unclear on whether the client is sending the input data block-by-block or as a single big chunk (which would rule out 3 for example). So who knows, it might be option 5. I am probably using the wrong language here anyway, as I'm aware that there is not a simple client/server separation in this case. A diagram might be useful!

I wrote a wiki page and drew a diagram about this subject, but in my case thinking about this as a network protocol: https://code.soundsoftware.ac.uk/projects/vamp-js/wiki/Client_Progress_Reporting_in_Higher-level_API (though I've just realised that the wiki page is private while this issue is public - perhaps something to be fixed later). The executive summary there (for a protocol) is that I think we need to draw a clear line and say that, if the server is going to do the reshape, i.e. if the client is calling the `collect` function, then the client _cannot_ receive features incrementally, _ever_. They always come back in a single shaped structure, not as a lazy sequence. That's probably OK for many client/server scenarios, especially where network traffic dominates, but it might not really be answering the right question here.

**Comment by cannam** [comment]
Meanwhile, another change to the structures in HigherLevelUtilities.ts. The main functional change is to give VectorFeature a start time, which means TracksFeature can simply be an array of those:

```TypeScript
export interface VectorFeature {
    startTime: number;
    stepDuration: number;
    data: Float32Array;
}

export interface MatrixFeature {
    startTime: number;
    stepDuration: number;
    data: Float32Array[];
}

export type TracksFeature = VectorFeature[];

export type FeatureCollectionShape = "matrix" | "vector" | "tracks" | "list";

export type FeatureCollection = {
    shape: FeatureCollectionShape;
    collected: VectorFeature | MatrixFeature | TracksFeature | FeatureList;
}
```

With this relationship, I think it becomes more natural to (revert to) making VectorFeature, MatrixFeature and TracksFeature singular nouns. Otherwise a tracks feature would be a VectorFeatures[] which is a bit weird. And it probably is more reasonable to think of a single vector/matrix as a feature rather than an aggregation of same.

**Comment by LucasThompson** [comment]
My current thinking aligns with most of the points in "Further notes", and some of those ideas have manifisted themselves concretely within the current mess of an implementation. 

> "The business of sending the audio data to be processed could actually just involve a single big HTTP request payload rather than the series of small blocks"

This has been the model since the birth of `HigherLevelUtilities`.

Before the changes to update a progress bar, this process was modelled with the `SimpleService` methods, so basically a function mapping `SimpleRequest -> SimpleResponse`. A `SimpleRequest` payload contains all the audio and metadata about the features to extract. `SimpleResponse`, at least when it was first defined, contains all the features summarised into one of the 'collected' shapes.

The additional work to power the progress bar used the above types and ultimately produced `StreamingService`. A `StreamingResponse` is just a `SimpleResonse` with some additional progress information. The server actually reshapes each individual `FeatureList` (a block of features for a single output, calculated from one call to process) and sends that to the client. This is, perhaps, where the confusion starts (and the bad design decisions). Amongst other things, this results in response payloads which are larger than necessary and which contain mostly redundant information (the shapes are deduced from the output descriptors, so they aren't going to change during the duration of the extraction / streaming process - so only need to be reported once really). 

The introduction of tracks screws this up, because the shape is deduced by looking at the data as it comes in and so can change mid stream. As mentioned in the first point of "further notes", attempting to handle this at a protocol level complicates things and is probably best avoided. 

At this point, it is appropriate to outline some implementation details of the current streaming extraction process:

- A client sends a single request with all the audio and metadata about the feature to extract
- The server extracts a single block, reshapes / collects it based solely on that output descriptor, adds some information about the number of blocks processed so far and total blocks to process and streams it to the client
- The client combines these features, ultimately into a single `FeatureCollection`, like what would've been returned in one go from the old `HigherLevelUtilities` `SimpleService`. 

In ugly-duckling, this looks like:

- feature-extraction.service.ts (client, uses a piper client internally - `WebWorkerStreamingClient` from piper-js). 
    - This is what the application uses to extract features. It uses an internal API, and isn't strictly a piper client. Progress events are emitted for a particular extraction, and features emitted at the end of the extraction. 
- FeatureExtractionWorker.ts (server). Internally, this is where things get nice and confusing and blurry lines between client and server appear. Layers and layers of indirection. 
    - `ThrottledReducingAggregateService` does the combining (uses functions from FeatureReducers.ts), keeping most of the heavy lifting on the worker side of the app
    - `ThrottledReducingAggregateService` combines streaming responses and in turn responds to the client after there is 1% change in progress, containing the combined features up to that point (which are never actually consumed / used in ugly-duckling) and the progress information (after all, it is just a `StreamingResponse`) (this was a bodge solution when the browser event queue was getting swamped with messages posted from the worker)
    
    
So the actual implementation is a mix of a bunch of your suggestions, but somewhat closest to 2 with a bit of 4 mixed in (and probably others). The server sends a response after each process call, reshaping that *single* block into a `FeatureCollection`. These are then assembled further, actually technically by the server and forwarded to the client at different points solely for the progress information (even though there is feature data in the response collected up until that point). By the last `StreamingResponse`, the server responds with all the collected data, which is then emitted by feature-extraction.service.ts and it updates the state of the application and it all falls out the bottom and gets added as a card etc. 

It seems sensible to me that this gets simplified to something like the following:

- Client sends something like a `SimpleRequest` with all the audio and metadata for the desired feature extraction
- The Server responds with raw features / "list" type stuff (we don't need the shape information) and progress information
- The client (even if this happens to be in the blurry land of the worker in ugly-duckling) assembles / reshapes the features as they come in

So using the same building blocks a server or client can handle the reshaping.

----

Also, some additional notes whilst I am busy dumping my brain into this box:

Most of HigherLevelUtilities isn't used, and I feel should just be deleted. We want to end up with a set of functions which can be combined to build servers or clients (and perhaps actually provide implementations of servers and clients - like the WebWorkerStreamingClient and WebWorkerStreamingServer)

The only methods used (and by used I mean in ugly-duckling, I realise this is supposed to be a general library for building piper clients / servers):

- `loadAndConfigure`
- `reshape` - this is the core function which needs splitting out / changing to be used in different scenarios 
- `segment`
- `toProcessInputStream`

**Comment by LucasThompson** [comment]
Also, whilst I think about it. For the purpose of ugly-duckling, and for when waveform.component.ts eventually gets some attention, it would be useful that by the time the features are emitted from feature-extraction.service.ts, they are emitted in a form which can directly be used to choose the right component to render. i.e. the business of deciding if it is a region, or notes or whatever happens there.

**Comment by cannam** [comment]
Thanks, this is interesting. I hadn't thought of that explanation.

I think there is another reason why this model would go wrong with the tracks feature -- you can't get a tracks feature back from one process block and naively merge it with another tracks feature from another process block, because you don't know whether the values in it are a continuation of the same track or belong to a new one. (You can guess using the same logic as is used to separate them into tracks in the first place, but why do the same work at both ends?)

(Later edit: and of course in the code as we have it now, where individual process blocks are individually reshaped, the blocks will always return vector type because there is never a gap between features within a single process block. So the reducer will always merge them into a single long vector. The tracks feature never becomes an option at all.)

Your suggested simplification appears to correspond roughly to option 1 from my earlier list. The server always returns raw list features, the progress update is based on the number of raw list feature sets returned, and the same reshape logic that happens on the server now would happen in the client instead. That would also be faster than what we have now, as it would avoid having to run the reshape logic for every process block.

The disadvantage of doing it that way is that it means we are talking only the low-level Piper protocol, and any putative feature extractors (or servers) that talked the higher-level protocol directly would need a different mechanism. Since quite soon we expect to have a server that does that, it's going to need different handling.

For a server, of course, progress updating can be based on the network transmission time rather than processing time (won't always be accurate but with current network technology would usually be better than nothing) which would fit OK with using the higher-level API with a totally separate handling of progress etc. However, that doesn't really fit if we're using the higher-level API in-process.

A perhaps-viable alternative would be more like option 3: something in the "middleware" layer (around the collect function in HigherLevelUtilities) accepts the audio input in one big chunk, but then explicitly provides progress updates back to the client during the procedure of chopping it up and feeding it to the extractor in process blocks. That would give exactly the same progress granularity as we get now. However, it wouldn't actually return any features until the collect function finished, when it would do a single reshape and return the lot. From the client point of view, this would just look like calling the existing collect function (none of the current explicit streaming stuff) but with an optional progress reporting callback.

What do you think?

Re. your last remark, yes, and that ties in also with #11 "can't represent notes" which is also a priority issue. At the very least, we could provide a way for extractors to present an accompanying ontology URI for extracted features. I should probably take a look at that one.

**Comment by LucasThompson** [comment]
I'm finding it quite hard to decide here. It strikes me that, with a little bit of work, by combining / re-jigging functions already implemented we essentially already have everything we need to support all these different variants. Though, from an api design point of view we are going to need to formally pick one. 

We want to support a number of different use cases. Whilst we aren't actually doing incremental rendering, we want to be able to support use cases like that. Of course that can just be done by using the low-level api, but the subtle difference here is in sending all the audio in one request and getting the feature blocks being streamed back.

When I bailed out of incremental rendering, I ended up with the ugly-duckling looking a bit like option 3, because it still only wants the collected features in a single response but to have progress updates throughout the process - but obviously ended up with a bunch of stuff going on in transmission that doesn't need to be (in this specific use case, although the payloads are wrong regardless).

Right now, something like option 3 is okay so that we basically end up with what we already have for use in ugly-duckling but with the ability to get tracks. This might be fine. That may be what we need for the server we are going to build soon too. I think the difference with my thoughts this whole time is that I want to support incremental rendering, and it'd be nice if the library building blocks were modular enough that you can assemble the processing pipeline you need a la carte. This is why I modelled the clients and servers with a single symmetric interface, `Service` (and the higher level `StreamingService` for example) so that you can set up these pipelines in largely similar ways regardless of whether combining of features happens on the server or the client. I may not have got it quite right yet, but I believe it still may be largely possible. I am not sure at this point how much these are just abstractions at the library level and what needs to be formalised in an actual api / protocol for use in the higher level server we intend to build soon.

**Comment by cannam** [comment]
Yes, I like the reasoning there. And it ought to be quite easy to refactor what we have now into the simplification you suggested (the "option-1-ish" version) and then rework the bits we have into something else when the server comes along.

**Comment by LucasThompson** [comment]
And with that, I deep dive back into piper-js.

**Comment by LucasThompson** [comment]
Okay, I've done the "option-1-ish" changes to the StreamingService api in piper-js and updated ugly-duckling accordingly, both on dev branches in my forks. Performance seems similar. 

Try checking them out or merging into yours and let me know.

I encountered that error again after building from a clean install of dependencies, that puzzling error originating somewhere in the worker bootstrap. It seems to happen *unless* I install with yarn with current yarn.lock which is in the repository, implying something has changed in a recent update of something (but I really have no idea). Maybe it'll fix itself, but I fear I may have to investigate this at some point.

**Comment by cannam** [comment]
Nice, thank you. I've merged and it's working here. Now I just need to get the tracks shape rendered appropriately!

**Comment by cannam** [comment]
The duckling now uses multiple line layers for the tracks features. We now get accurate visualisation for pYin pitch tracks, which I have to say is f*ing ace. However, I haven't yet done the work to get the highlight layer to span all the tracks, so we're not quite finished yet. (The highlight layer is simply disabled at the moment.)

**Comment by LucasThompson** [comment]
Very cool! Now we need to make it quack! Sonify your melody lines as a duck! ([inspiration](https://www.youtube.com/watch?v=0c-CvPgyO6s))

**Comment by cannam** [comment]
The highlight layer is reinstated now (in my dev/tracks-shape branch and with corresponding waves-ui-piper update). It doesn't scale all that well to many tracks, but let's give it some testing with real data before we decide whether to worry about that.

Let me know whether all this is working for you.

**Comment by cannam** [comment]
Looks like it works fine, but it's quite slow to update (i.e. refresh the layers during normal use) if there are many tracks. Have to have a look and see if this can be improved.

**Comment by cannam** [comment]
Have made some improvements (yesterday, in fact) to the refresh logic and it's a bit faster now. Closing this one! Woo!
